bpf/lib/l4.h

```diff
@@ -135,26 +135,51 @@ struct l4_allow
 	__u8 nexthdr;
 };
 
-#if (defined CFG_L4_INGRESS || defined CFG_L4_EGRESS) && !defined CONNTRACK
-#error "CFG_L4_* requires CONNTRACK to be enabled"
+#if (defined CFG_L4_INGRESS || defined CFG_L4_EGRESS || \
+     defined CFG_L3L4__INGRESS || defined CFG_L3L4__EGRESS) && !defined CONNTRACK
+#error "CFG_L* requires CONNTRACK to be enabled"
 #endif
 
-#ifdef CFG_L4_INGRESS
-static inline int __inline__ l4_ingress_embedded(__be16 dport, __u8 nexthdr)
+#if defined CFG_L4_INGRESS || defined CFG_L3L4_INGRESS
+static inline int __inline__ l4_ingress_embedded(__be16 dport, __u8 nexthdr,
+						 bool l3dependent)
 {
 	int allowed = DROP_POLICY_L4;
 
+#ifdef CFG_L4_INGRESS
 	BPF_L4_MAP(allowed, dport, nexthdr, CFG_L4_INGRESS);
+#else
+	/* If there are only l3-dependent L4 rules, then CIDR-based L4 is
+	 * completely wildcarded */
+	allowed = 0;
+#endif /* CFG_L4_INGRESS */
+#ifdef CFG_L3L4__INGRESS
+	if (l3dependent) {
+		BPF_L4_MAP(allowed, dport, nexthdr, CFG_L3L4__INGRESS);
+	}
+#endif /* CFG_L3L4__INGRESS */
 	return allowed;
 }
 #endif
 
-#ifdef CFG_L4_EGRESS
-static inline int __inline__ l4_egress_embedded(__be16 dport, __u8 nexthdr)
+#if defined CFG_L4_EGRESS || defined CFG_L3L4_EGRESS
+static inline int __inline__ l4_egress_embedded(__be16 dport, __u8 nexthdr,
+						bool l3dependent)
 {
 	int allowed = DROP_POLICY_L4;
 
+#ifdef CFG_L4_EGRESS
 	BPF_L4_MAP(allowed, dport, nexthdr, CFG_L4_EGRESS);
+#else
+	/* If there are only l3-dependent L4 rules, then CIDR-based L4 is
+	 * completely wildcarded */
+	allowed = 0;
+#endif /* CFG_L4_EGRESS */
+#ifdef CFG_L3L4__EGRESS
+	if (l3dependent) {
```

Comment:
Let's reset allowed and treat each call to `BPF_L4_MAP()` independently.

Reply:
Done, separated out into entirely different invocations.